credential-manager/src/CredentialManager/Scripts/Common.hs

```diff
@@ -10,28 +11,209 @@
+                Nothing -> trace "Invalid redeemer" False
+                Just redeemer ->
+                  checkSpend
+                    scriptContextTxInfo
+                    ownRef
+                    txOutAddress
+                    txOutValue
+                    datum
+                    redeemer
+          | otherwise = go inputs'
+    _ -> trace "Not a spending script" False
+
+{-# INLINEABLE checkContinuingTx #-}
+checkContinuingTx
+  :: (FromData a) => Address -> Value -> [TxOut] -> (a -> Bool) -> Bool
+checkContinuingTx addrIn valueIn outputs checkDatum =
+  case findOutputsByCredential (addressCredential addrIn) outputs of
+    [TxOut addrOut valueOut datumOut _] ->
+      traceIfFalse "own address not preserved" (addrIn == addrOut)
+        && traceIfFalse "own value not preserved" (valueIn == valueOut)
+        && case datumOut of
+          OutputDatum (Datum datum) -> case fromBuiltinData datum of
+            Just datum' -> checkDatum datum'
+            Nothing -> trace "Invalid output datum" False
+          _ -> trace "Output has no datum" False
+    [] -> trace "No continuing output found" False
+    _ -> trace "Multiple continuing outputs found" False
+
+{-# INLINEABLE checkResignation #-}
+checkResignation
+  :: [PubKeyHash]
+  -> Identity
+  -> (a -> [Identity])
+  -> a
+  -> a
+  -> Bool
+checkResignation signatories resignee getGroup datumIn datumOut =
+  traceIfFalse "resignee is not a group member" checkGroupMembership
+    && traceIfFalse "resignee has not signed transaction" checkSignature
+    && traceIfFalse "resignee is last member of the group" checkNonEmptyOutputGroup
+    && traceIfFalse "unexpected group in output" checkOutputGroup
+  where
+    oldGroup = getGroup datumIn
+    newGroup = getGroup datumOut
+    checkGroupMembership = resignee `elem` oldGroup
+    checkSignature = pubKeyHash resignee `elem` signatories
+    expectedNewGroup = filter (/= resignee) oldGroup
+    checkNonEmptyOutputGroup = not $ null expectedNewGroup
+    checkOutputGroup = newGroup == expectedNewGroup
+
+{-# INLINEABLE checkRotation #-}
+checkRotation
+  :: [PubKeyHash]
+  -> (a -> [Identity])
+  -> a
+  -> a
+  -> Bool
+checkRotation signatories getGroup datumIn datumOut =
+  traceIfFalse "New group empty" checkNonEmptyOutputGroup
+    && traceIfFalse "Added user signatures missing" checkSignatures
   where
-    toSelf (TxOut address _ _ _) =
-      addressCredential (txOutAddress ownInput) == addressCredential address
-
-{-# INLINEABLE checkTxOutPreservation #-}
-checkTxOutPreservation :: TxOut -> [TxOut] -> Bool
-checkTxOutPreservation ownInput outputs = case ownOutputs ownInput outputs of
-  [output] -> output == ownInput
-  _ -> False
+    oldGroup = getGroup datumIn
+    newGroup = getGroup datumOut
+    checkNonEmptyOutputGroup = not $ null newGroup
+    checkSignatures = all signedIfNew newGroup
+    signedIfNew i@Identity{..} =
+      i `elem` oldGroup || pubKeyHash `elem` signatories
```

Comment:
This is a very nice check! Prevents unwitnessed users :heart:

Reply:
Do you think it should go in the minting validator checks too?